### ISSUE TITLE
Intuition about hyper parameters for Atari games

### ISSUE THREAD

**Issue by piojanu** [google-research/planet#21]
Hi!

I decided to put Sokoban off for now as it doesn't seem to work the same as in World Models. So I started to experiment with Atari once again. I easily made World Models work with e.g. Boxing environment, but in PlaNet I get very blurry open loop predictions:
![image](https://user-images.githubusercontent.com/23562932/55856042-6672bd00-5b69-11e9-8dc3-07d49166dcd6.png)
Please, could you share your intuition about hyper-parameters I might try to tune? I tried to disable overshooting (to make the case more similar to World Models) and now I'm looking into divergence scales, but without luck for now. As I have no cluster to run hyper-params optimisation, maybe you have some ideas what is worth focusing at? :D

Thanks!

**Comment by Kwander** [comment]
At least for me, 50k train, 1.5k test and collect_every 8k works for Assault-V0 with 1.2k max length bound. The agent learns somewhat decent behaviour (~300 score per episode). Note that Atari games already do action repeat, thus I disabled this in the PlaNet's source code.

**Comment by astronautas** [comment]
@piojanu The predictions don't seem that blurry (I do see agents fighting). Take a look again into World Models paper - their predictions look blurry as well.

**Comment by piojanu** [comment]
Thanks guys, I still work on that. Nevertheless, PlaNet continuous control tasks had really sharp reconstructions even after 15 steps. Furthermore, I've implemented World Models and trained it for Boxing. Here are the open loop predictions I was able to obtain there:
![image](https://user-images.githubusercontent.com/23562932/55906847-7aec9f00-5bd5-11e9-9083-89ea4be30b52.png)
As you can see those are MUCH clearer. Do you have other ideas?

**Comment by astronautas** [comment]
@piojanu Interesting, they do seem much clearer. That's a dimension to explore though, maybe the VAE of World Model's is bit better :)

**Comment by piojanu** [comment]
Also:
@astronautas the agent performance is weak too. Please keep your responses more succinct. What do you mean "maybe the VAE of World Model's is bit better"?
@Kwander thank you for noting that OpenAI Gym already does action skipping. I'll disable it too!

**Comment by piojanu** [comment]
Also:
@astronautas the agent performance is weak too. Please keep your responses more 
@Kwander thank you for noting that OpenAI Gym already does action skipping. I'll disable it too!

**Comment by piojanu** [comment]
Closed by mistake, sorry.

Lowering global prior divergence scale seems to help. I'll disable it and run the test again ;)

**Comment by astronautas** [comment]
@piojanu What do you mean the performance is weak too? I've edited my answer.

**Comment by piojanu** [comment]
He gets on average below 0 points. In World Models it was 18 points on average, but planning algorithm is different so it's not fair to compare. Please lets focus on prediction metrics (reconstruction in that case) as it's the main focus of my question.

**Comment by astronautas** [comment]
What's the batch you're using?

**Comment by piojanu** [comment]
@astronautas that's funny a little, expanding VAE to Variational Auto Encoder doesn't add value I was thinking about 😄
I use default batch size (50, 50).

**Comment by astronautas** [comment]
Sorry, the author did not use the term vaes in the paper, Variational encoder was the definition used :D

**Comment by piojanu** [comment]
@Kwander do you observe this too:
<img width="877" alt="image" src="https://user-images.githubusercontent.com/23562932/55908390-a07ba780-5bd9-11e9-8f22-b41b761a1eb7.png">
Please note the scale of y-axis. Those values are barely changing e.g. `graph/summaries/closedloop/posterior/log_prob/image` stays pretty much the same all the time. Is it expected? For me it's sign that it's not learning much beyond first steps before first summary gets written.

**Comment by astronautas** [comment]
@piojanu I have noticed this as well, for dmcontrol cartpole swingup problem.

**Comment by piojanu** [comment]
Okay, thanks for clarification ;) Do you use default config in "Assault-V0"? Can you post your open loop reconstructions and agent's score to see what to expect?

**Comment by danijar** [comment]
@piojanu By default we use a decoder variance of 1, which means the model explains a lot of variation in the image as random noise. While this leads to more robust representations, it also leads to more blurry images. If the predicted images are all the same, the posterior collapsed because the model explains everything as observations noise. Try to reduce the decoder variance in conv_ha.py or equivalently set a lower divergence_scale parameter.

@astronautas That's a cool result. I assume you're still using CEM with a Gaussian belief over action sequences for planning on even though the Atari actions are discrete, and later take the argmax along the action dimension?

**Comment by astronautas** [comment]
@danijar Thanks, yes, I am using those. My action space consists of N actions (-1 to 1), which I later argmax in the step function of the environment wrapper.

**Comment by piojanu** [comment]
Thanks guys, status update:
For smaller state size (16) and smaller global divergence scale (0.01) I got to this:
![image](https://user-images.githubusercontent.com/23562932/56028027-82b45c80-5d17-11e9-8233-6f7b4789d2c2.png)
Today I'll run hyper-param tuning and leave it for ~3-4 days. I'll get back to you with results (and even more questions, haha 😊) in here so let's leave this issue open.

**Comment by piojanu** [comment]
Quick update: I'm working on the hyper-param tuning. I wasn't able to run it during the last weekend. I run it yesterday though.

**Comment by astronautas** [comment]
I did several more runs on Assault and Breakout Atari. I'll update my comment in the evening with the results. The algorithm seems to perform worse on Breakout game, although it's probably the simplest one for model-free algorithms (no early conclusions, will share the results first).

**Comment by piojanu** [comment]
@danijar check me if I understand correctly what is happening.
1. Consecutive frames in Boxing are quite similar so the decoder explains it as a random noise. It means that the posterior starts to infer similar codes for consecutive frames (this is what you mean by "the posterior collapse", right?).
2. The transition model (prior) is trained to minimise KL-divergence to the posterior. The posterior tells it that consecutive frames have (almost) the same codes, so the transition model doesn't learn anything and open loop predictions are more or less the same blurry blob.

Now, lowering the variance of the decoder should result in sharper reconstructions, because a blurry blob is less probable under the lower variance i.e. MSE loss is higher for blurry recostructions and the posterior needs to encode concrete information about what is in the frame so the decoder can use it (i.e. it needs to stop encoding bull**** 😄).

The other way is to lower the divergence scale, so one-, two-, three-... step predictions (that will have the same or very similar codes as I explained above) will have lower impact than long-term predictions (which should differentiate the frames better, because frames further apart have more differences). Am I right? 

One thing that I don't understand is why you called the two approaches "equivalent":
> Try to reduce the decoder variance in conv_ha.py or equivalently set a lower divergence_scale parameter.
Those seems to me to address the problem in fundamentally different ways and maybe both changes will need to be applied. Could you expand on why those are equivalent?

**Comment by danijar** [comment]
I suggest to reduce the divergence_scale (I'd try values around `1e-2` or `1e-3`) and to increase the action repeat. The action repeat will result in a bigger difference between consecutive frames and thus more signal for the model to learn from, that cannot easily modeled as noise as you explained.

Divergence scale and the (constant and scalar) decoder variance are the same. You can see this by writing the ELBO for a Gaussian decoder in the standard form `E_q(z)[lnp(x|z)]-KL[q(z) || p(z)]`. The log-likelihood terms is `lnp(x|z) = -0.5(x-f(z))/std^2-lnZ`. Multiplying the ELBO by `std^2` removes it from the log-prob term and puts it in front of the KL term as in beta-VAE. The objectives have different values because of the Gaussian normalizer `Z` but they have the share the same gradient since the normalizer is a constant.

**Comment by astronautas** [comment]
So here's some result when training to play Breakout-v0:
- Episodes: 170
- Average test performance: 4
- Action repeat (probability): 0.25 <- Atari default

Sadly, I accidentally deleted the log directory, thus I cannot share images. Yet, what I noticed was that the ball was not visible in graph/summaries/closedloop/posterior/image/animation/gif. I'll try running some experiments with @danijar suggested divergence rate.

Note: I am mapping [-1, 1] interval to [0, action_space.n] and doing math.floor to get a discrete action. E.g. if the network predicts -0.1, it gets mapped ~2.95, which gets transformed into action 2.

Things yet to try:
- Sampling actions from a discrete Gaussian in planning to avoid the mapping above.
- Trying out different divergence rates, as suggested.

**Comment by astronautas** [comment]
EDIT: I did another run with this configuration:
    params.batch_shape = [24, 50]
    params.train_steps = 40000
    params.test_steps = 1000
    params.collect_every = 20000
   + default
   
* Max_length: 1200, maximum_episode_duration: 600.

![graph/summaries/closedloop/posterior/image/prediction/image/0](https://i.imgur.com/nSwKqfh.png)

**Comment by piojanu** [comment]
@astronautas Here you have the graphs from training simple [VAE on MNIST dataset](https://github.com/piojanu/tf_utils/tree/master/samples) (I've implemented it in TensorFlow Probability using Danijar coding patterns from his blog to better understand PlaNet summaries and codebase ~and take over his researcher identity~. I start to feel like some creep/crazy fanboy 😮 @danijar do you or PlaNet have fanpage? haha 😄):

![image](https://user-images.githubusercontent.com/23562932/56273259-b408a000-60fc-11e9-9d96-b0b6ba81710e.png)
![image](https://user-images.githubusercontent.com/23562932/56273339-d4d0f580-60fc-11e9-91a1-73f373455263.png)
![image](https://user-images.githubusercontent.com/23562932/56273387-eca87980-60fc-11e9-92d6-fc42a1b7180f.png)

So, let's get serious, my intuition is that KL-divergence from a global prior (which is a normal distribution) should get higher as the posterior is trained to encode useful information. At the same time, it makes an entropy of the posterior lower (it's more and more certain about the latent state (code) as it learns). On the other hand, KL-divergence between the posterior and a prior from a transition model should get lower as the transition model learns to predict future latent states (codes) more and more accurately. loss (negative ELBO) needs to get lower, but all the log prob. needs to get higher.
@danijar please confirm that it holds for PlaNet. Also, I still don't understand if it's correct that summaries has so small ranges in y-axis (see my comment [above](https://github.com/google-research/planet/issues/21#issuecomment-481835393)).
@danijar also thank you for your explanation about beta-VAE, I didn't catch that.

Also, @astronautas please post `openloop` and `closedloop/prior` predictions as those are interesting too (or even more interesting). `closedloop/posterior` in 0-step prediction (encode and decode the frame), `closedloop/prior` is 1-step future prediction (+ decode) and `openloop` is n-step future prediction (+ decode). I start to gather interesting results from my hyper-param tuning for Boxing. I'll post the results have tomorrow or next week after Easter holidays.

**Comment by astronautas** [comment]
Here's some results from my latest run (divergence scale 1e-4, learning rate 0.5e-3)

Openloop:
![openloop](https://imgur.com/4e6174b2-1014-40f4-908a-b77a0e79c53f)

Closedloop posterior:
![closedlooprior](https://imgur.com/8fb9ced9-1d0a-4ce5-83dd-3d2dd4e4505f)

Closedloop prior:
![closedlooprior](https://i.imgur.com/mz47xL0.png)

Entropy:
![entropy](https://imgur.com/faaf8855-6e28-4c51-a810-f7ac2054e4df)

Correct if I am wrong guys but the posterior entropy should be decreasing, and log_probs should all be increasing, am I right?

Thanks!

**Comment by danijar** [comment]
The entropy and losses don't necessarily have to decrease during training, because the agent is collecting new data that could be surprising. I would mostly look at the openloop images summaries and the predicted reward trajectories (another image summary).

@piojanu I don't have a fan page, unfortunately :D I'm not sure if your intuition is correct regarding the KL. the KL to the global prior may be roughly constant throughout training.

**Comment by astronautas** [comment]
@danijar Good tip, here are the predicted reward screenshots:

![rewards](https://i.imgur.com/UKEmAvc.png)

It makes sense why the performance is so poor - the planner is not capable of successfully predicting reward spikes. Concerning the openloop, how should one interpret them? Does it follow the same structure as it was in the paper:

![openloop](https://i.imgur.com/9lsaT5L.png)

**Comment by danijar** [comment]
Yes, the open loop images have 5 context frames and all other frames are predicted open loop. For the paper, we additionally only showed every 5th frame to have it fit on a page. The TensorBoard summaries don't skip any frames.

It seems like the reward predictions are not very accurate (yet). How long did this train for? I would expect it to take at least a couple of million steps. One million steps means 200 episodes since it collects an episode every 5000 steps.

**Comment by astronautas** [comment]
Cool, thanks for clarifying this!

I trained it for about 6 hours and 800k steps (370 episodes). 

The problem with these Atari games is that more than one episode is collected in one collect session, as the "done" signal is issued when the agent loses all lives. I was thinking maybe of writing a wrapper to reset the game when the agent runs out of lives to solve this or collect data till first done is fired.

**Comment by piojanu** [comment]
@danijar could you share your insight about KL-divergence to global prior then?
@astronautas I think episodes are reset after "done", it's here:
https://github.com/google-research/planet/blob/9cd9abed5b9a8831388f4d9da16e5604cfbd7c20/planet/control/simulate.py#L225-L230
@danijar could you confirm that?

**Comment by astronautas** [comment]
@danijar @piojanu I have some new Breakout results after fiddling with the divergence_scale.

It seems it's going be tough yet interesting. With high divergence scale, ball 0 step reconstructions fail. With the reduced rate to 1e-4, it mostly succeeds at 0 step reconstruction, though hallucinates incorrect future ball positions or incorrectly infers ball movement direction.

Sometimes it does succeed. Last row of the image below - in the 6th frame, it confidently chooses a corrective action. Still, it is kind of "shy" about predicting the ball movement.

![openloop](https://i.imgur.com/ktIEbVK.png)

**Comment by danijar** [comment]
The fact that the ball is only a few pixels large might be a problem, with a loss that treats all pixels equally there is little motivation for the model to learn to predict it -- it's much easier to get the loss down by getting really good at predicting the background. Lowering the divergence scale should help a bit as you said. You can also try to reduce the color depth further, e.g. to `image_bits: 3`. At the same time, it might be easier to start with an environment that doesn't have important but tiny objects.

**Comment by astronautas** [comment]
Thanks @danijar for the tips. I am still eager to maximize PlaNet's accuracy on Breakout - lots of games contain small yet important details which should not be ignored.

I am intrigued by the World Model's @piojanu shared. This probably suggests that it's worth fiddling with the loss functions as well. Thanks for keeping this discussion alive. Hopefully, this will provide some insights to you as well about what PlaNet still lacks.

**Comment by danijar** [comment]
@astronautas The `global_divergence_scale` is for the fixed global prior and the `divergence_scale` is for the temporal prior predicted by the model. The prior and posterior are the same as described in the paper, prior means `p(s_t | s_t-1, a_t-1)` and posterior means `q(s_t | s_t-1, a_t-1, o_t)`. The state `s_t` is a dict that contains both the deterministic GRU state and the stochastic state.

I've found that I can train on the MuJoCo tasks without overshooting, global prior, and reward scale, by setting `future_rnn: True` and switching the activation functions from `tf.nn.relu` to `tf.nn.elu`. We'll post an update to the paper at some point but if you like you could already try and see if this works better on Atari as well. Another thing to try to help with small objects is to set `free_bits` to a larger value.

**Comment by astronautas** [comment]
@danijar Thanks! If we disable overshooting though, does the loss still include training of priors against posteriors (i.e. KL (p(s_t | s_t-1, a_t-1); q(s_t | s_t-1, a_t-1, o_t)) ? It looks like this is disabled as we disable latent overshooting loss, though correct me if I am wrong.

**Comment by danijar** [comment]
There will still be a KL term from each posterior to the temporal prior predicted from one step before. This is the term from the standard ELBO that is the only term left when setting `overshooting: 0` and its scale is controlled by `divergence_scale`.

**Comment by astronautas** [comment]
Thanks for clarifying this @danijar. I got confused, as in the paper it seemed to be included in the [3rd](https://i.imgur.com/EDPgWpq.png) equation, yet it wasn't there anymore in the final [7th](https://i.imgur.com/bo0di3h.png) equation. But maybe it's just me ¯\_(ツ)_/¯.

**Comment by danijar** [comment]
Exactly, with D=1 latent overshooting reduces to the standard ELBO. It's just a bit confusing that the `overshooting` parameter in the code is not `D` and `D-1`, such that `overshooting: 0` means standard ELBO.

**Comment by piojanu** [comment]
What is `future_rnn: True` and `free_bits`?

**Comment by astronautas** [comment]
@danijar I haven't found _free_bits_ parameter in the source code as well.

**Comment by danijar** [comment]
Sorry, it's called `free_nats` and means the model is allowed to use this amount of nats without KL penalty, a trick that's often used for static VAEs. It helps the model focus on smaller details which don't contribute much to improving the reconstruction loss. The `future_rnn` flag fixes a somewhat somewhat subtle bug in the RSSM code, where RNN and stochastic state were both used but didn't interact with each other at future steps.

**Comment by astronautas** [comment]
@danijar Thanks for the info, I'll try these!

I am also curious, what's the use of the observation model? I've read it's only used for training as a "signal", which intuitively sounds valid, as we want latent states to contain visual information. Could we though train only the reward model, so that the latent states would be more "focused" on encoding the reward relevant visual information? 

TLDR; What was the motivation behind including the observation model for training although it's not used for planning?

**Comment by danijar** [comment]
I've experimented with learning only reward models a while ago. This works in surprisingly many cases, however, it doesn't work well when there are sparse rewards, as in cup catch. Moreover, it's a bit less data efficient, but still much more data efficient than most model-free approaches. I think the effect of the observation model is comparable to the effect of auxiliary losses in UNREAL, etc. But I also want to move towards agents that learn without reward, which makes predicting sensory inputs more important.

**Comment by Kaixhin** [comment]
@danijar thanks for all your hard work open-sourcing this. I'm working on a PyTorch port and it was fundamental in working out the architecture (even though the paper was very clear about most other aspects). Just to confirm, your reported results were without fixing the `future_rnn` bug? I've managed to replicate your results (i.e., overshooting is important) having fixed that in my code on the `walker-walk` environment, so any comments of the importance of that vs. replacing all ReLUs with ELUs?

FYI don't think I spotted any other oddities in your code. Looking to do some research on top of PlaNet, hope to open source my code once done. Unfortunately I'm looking at almost 2 weeks per run atm and a not-insignificant amount of memory usage, so will have to look into possible speedups, but the algorithm in general is quite sequential so not sure much can be done.

**Comment by danijar** [comment]
@Kaixhin That's great! The results where without the fixed bug and I'll update them in the paper. There are a few other small updates to the paper that I want to make (add some citations, etc), so it might take me a couple of days. I've confirmed that elu is not the deciding factor.

I'd be great to see some plots comparing with and without overshooting using your implementation, if you have them. Maybe you could open a new thread for everything related to your PyTorch rewrite?

**Comment by danijar** [comment]
Hi @astronautas, a few updates from my side:

- It turns out when I added the `future_rnn` parameter in the open source repo, I passed it to SSM instead of RSSM in configs.py. This is fixed now.
- The updated agent uses tfd.Normal distributions instead of the custom MSEDistribution I used before in networks/basic.py and networks/conv_ha.py.
- I've also confirmed that ELU is not important so I've switched back to ReLU. It only made a different for a follow-up project I've worked on recently.

I think this could be worth another try for Breakout. I'll also sync all my changes to Github at some point for the camera-ready version of the ICML paper, but it could take a few weeks until I get to it.

**Comment by piojanu** [comment]
@danijar thanks for your update! I need to catch up with your posts. However, here are my results from random search of overshooting (from 0 to 50), divergance_scale (from 1E-5 to 1E-2) and global_divergence_scale (from 1E-5 to 1E-2) parameters. Every data point is from Freeway training for 1M steps. Those three scores are last three logged test scores in TensorBoard:
https://docs.google.com/spreadsheets/d/1VCQRyFVZzzbSE_LPpNIvu_QnzFSx9tV8g99mfVaK7gw/edit?usp=sharing
Well... for me it's some random noise :/ It seems like really low overshooting makes things worse, but I have other experiment where with no overshooting and global divergence set to 0.01 it does much much better (score ~16) after ~2M steps of training so it seems like random chance.
My conclusion is it doesn't matter that much if those scales are 1E-3 or 1E-5. What I can see clearly is that lowering those scales from default in deed helps. Also overshooting doesn't seem to have obvious advantage as well. @astronautas and @danijar could you look at this and tell what do you thing?

**Comment by astronautas** [comment]
@danijar I'll give it a shot with the new configuration on VizDoom, as I've wanted to do this from the very beginning. Thanks for the fixes!

@piojanu Thanks for sharing the results! Looks like big overshooting coupled with small global prior is needed, coupled with somewhat low, but not too low divergence scale. Enforcing consistency yet weak one for predictions of multiple steps seems to have helped, as well as reducing the dependency on forcing models to be Gaussian-like.

**Comment by piojanu** [comment]
I've also run training with `future_rnn: true, divergence_scale: 0.0001, global_divergence_scale: 0.0001` parameters for Boxing and Freeway. I intent to leave it to train for longer time. We will see what will turn out from that ;)

@astronautas what do you mean by "as well as reducing the dependency on forcing models to be Gaussian-like"?

**Comment by danijar** [comment]
I don't think lowering the divergence scale (normal or global) makes anything more or less Gaussian. The posteriors are conditionally Gaussian already. Note that the trajectory distribution is highly non-Gaussian because of the non-linear transformation at each time step. So starting from a Gaussian, the next time step is already a mixture of infinitely many Gaussians. One reason that lowering the divergence scale can help is that it allows the model to absorb more information from its observations by loosening the information bottleneck.

**Comment by astronautas** [comment]
@danijar Yup you're right, it makes sense :). 

One question - in Atari games, it crucial that the agent does not learn it's good to die to get a reward after environment is resetted. Are the episodes concatenated during training, or is the training done for each episode separately? Could be there training chunks containing <states,action,reward> tuples from different episodes?

**Comment by danijar** [comment]
All episodes are trained separately in the code, so as long as the episode terminates when the agent dies, this is not a problem.

**Comment by piojanu** [comment]
I'm back with good news :)

### TL;DR
Atari Freeway and Crazy Climber seems to start working. I get sharp open loop reconstructions, but still with some random "glimmer" of the chicken in case of Freeway and the building in case of Crazy Climber. Boxing also starts to work (posterior doesn't collapse any more)!

### Details

Hyper-params (if not stated otherwise):
- `global_divergence_scale: 0.0001`,
- `divergence_scale: 0.001`,
- `future_rnn: true` - this helped a lot!,

#### Boxing
**Experiments:**
I've run those for ~4.3M steps which is around ~870 episodes of Boxing.
1. `action_repeat: 8` - this didn't help with collapsing posterior, only made agent worse. Final score is below 0 which is worse than random agent which scores around 0. Before this higher action repeat experiment it is around 0 too, so it acts like random agent).
2. `free_nats: 4` and action repeat from above - higher `free_nats` helped a lot with collapsing posterior! Final score like above.

**Openloop and rewards predictions:**
![image](https://user-images.githubusercontent.com/23562932/56792484-00916100-680a-11e9-8be8-594149f25a66.png)
It still doesn't look great, but at least boxers aren't turning into blurry blob!

![image](https://user-images.githubusercontent.com/23562932/56792525-1b63d580-680a-11e9-9e97-5ded1a7265a5.png)
Rewards prediction doesn't seem to work. Any ideas how to improve it @danijar and @astronautas?

**Next actions:** I've run it again with standard `action_repeat: 4` and `free_nats: 4` for longer.

#### Freeway
**Experiments:**
Hyper-params listed at the top. Here I've verified that `future_rnn: true` helps a lot. I've run it for ~4.2M steps or ~850 Freeway episodes.

**Openloop and rewards predictions:**
![image](https://user-images.githubusercontent.com/23562932/56792931-0f2c4800-680b-11e9-8667-24284a89afd7.png)
In gif animation it can be seen even better that cars are precisely modelled and sharply reconstructed. However, chicken is reconstructed somehow randomly up and down. This needs to be improved.

![image](https://user-images.githubusercontent.com/23562932/56792924-089dd080-680b-11e9-8e21-0b043837e9f5.png)
Well, here rewards are really sparse, so there is not much signal to learn from to predict rewards.

**Next actions:** I've run it with `free_nats: 3` with hope that chicken will be modelled better. Also there is problem with rewards prediction to solve, any ideas guys?

#### Crazy Climber
**Experiments:**
Again, hyper-params from the top. I've run it for ~3M steps which is ~620 Crazy Climber episodes.

**Openloop and rewards predictions:**
![image](https://user-images.githubusercontent.com/23562932/56793814-1a807300-680d-11e9-9880-eff00c73bf6b.png)
Reconstructions are sharp, but building structure isn't modelled well (e.g. this black hole in the middle usually is much higher/lower then reconstructed). It can be seen better in gif.

![image](https://user-images.githubusercontent.com/23562932/56793570-9b8b3a80-680c-11e9-9217-70840190e023.png)
Again, it didn't train to predict rewards right.

Also agent have problems with moving up (some combination of actions is needed to raise a hand and then grab). Random agent does better. It seems like problem with exploration, what do you think?

**Next actions:** I've run it with disabled global prior to check if it changes anything. @danijar found out that it isn't important with fixed RSSM, see #28.

### Conclusions
I suspect that rewards aren't predicted too good and that's why it does so purely (in case od Crazy Climber it does worse them random, in [this paper](https://arxiv.org/abs/1903.00374) they have nice benchmarks and comparisons e.g. table 1. Their algo seems great too).
Other possibility, which I'll check, is poor exploration. We also need to remember that CEM isn't designed for discrete actio-space and maybe hacky way of argmaxing sampled action scores isn't the best idea (maybe we should use something softer, e.g. sampling from proportional/softmax dist. of those action scores?).

Im interested in what do you thing @astronautas and @danijar? **What next steps can you see?**

**Comment by piojanu** [comment]
Quick updates:
1. I've implemented this CEM modification to discrete action-space (one-hot actions + e-greedy exploratory policy in `mpc_agent.py`, not additive noise as in continuous action-space case).
2. `free_nats=3` helped with Freeway! Although there are still some errors in prediction, chicken moves are smooth and stable (don't break env dynamics with e.g. teleportation). I'm truly amazed right now and I start to believe it might work 😄
![image](https://user-images.githubusercontent.com/23562932/56850887-d1154e00-6908-11e9-9789-78fe42f1dd2e.png)

I'll post more probably tomorrow.

**Comment by astronautas** [comment]
@piojanu Did you try running same experiment with and without the discrete planner? Does it really improve the results?

**Comment by piojanu** [comment]
Sorry, I've edited my post. I'm now testing it ;)

**Comment by piojanu** [comment]
I have new results and I hope for your insights @astronautas and @danijar.

### Freeway - discrete planner experiment
Well, the agent have hard time getting to the other side of the road (where it gets a reward of 1). I've issued multiple runs with epsilon in e-greedy policy 0.3 and 0.7 and in three from four runs it didn't collected any reward (in data collection phase and in test phase):
![image](https://user-images.githubusercontent.com/23562932/56901157-c50bc680-6a97-11e9-96e5-bfda35691168.png)
It's even worse than without this discrete planner, I'll try to find out why. It it comes to modeling, then results aren't clear too. In 2 runs it models the actions really well in other 2 runs the chicken is jumping unpredictably. So I'll now try with disabling global prior.
Do you have some thoughts?

### Boxing - higher free nats and lower divergence scale
Still fighting, currently I tested divergence scales 1E-4 and 1E-5 with free nats 4 in multiple runs. Reconstructions are still looking like some randomly jumping blobs, EVEN for closedloop prior reconstructions (one-step prediction) in case of divergence scale 1E-5. So lowering the divergence scale further doesn't seem to make much sense. I'll try with higher free nats (5).
It's worth noting that closedloop posteriors looks fine, so it's still some problem with the transition model I bet. @danijar do you have some ideas?
Here are closedloop predictions for 1E-4 and then 1E-5:
![image](https://user-images.githubusercontent.com/23562932/56900597-7e699c80-6a96-11e9-96a9-c081dfa68843.png)
![image](https://user-images.githubusercontent.com/23562932/56900612-86c1d780-6a96-11e9-88f3-90506be19db5.png)

And here are openloop predictions for 1E-4 and then 1E-5:
![image](https://user-images.githubusercontent.com/23562932/56900830-15ceef80-6a97-11e9-823b-0b5b9da4cdda.png)
![image](https://user-images.githubusercontent.com/23562932/56900846-1b2c3a00-6a97-11e9-8c97-70407a7cb3a2.png)

### Crazy Climber - higher free nats
Nothing new, reconstructions are still not precise. I wan't post anything more as it looks much the same as in Boxing case. There are little differences between the frames and the transition model doesn't seem to get that. @astronautas any ideas how we can work that out?

**Comment by piojanu** [comment]
I made Boxing work really well! :D Score is high and open loop predictions are "sharp".

![image](https://user-images.githubusercontent.com/23562932/57614617-fe553380-7579-11e9-838c-f1657acb51ce.png)
![image](https://user-images.githubusercontent.com/23562932/57614853-8e937880-757a-11e9-98f6-dfc78c3b4985.png)

**TL;DR** ~7M steps; disabled global prior; enabled future rnn; used "discrete" CEM described above; divergence scale 3E-02; free nats 12;

# Hyper-params
The most important parameters seems to be free nats and divergence scale (and turning on future rnn). I did some tuning with many parameters and no other seem to have that big impact (I didn't try to change reward scales though). Random search results are: the lower divergence scale the nosier predictions are. The higher free nats the better actions movement predictions are (more stable one could call). Although, there were quite good reconstructions for low free nats too (when divergence scale was a bit higher or equal 1E-02). I'll try it next.

# Questions
@danijar what might cause so dramatic leap in image loss/log_prob statistics around step 2M? See the diagrams:
![image](https://user-images.githubusercontent.com/23562932/57613914-4ffcbe80-7578-11e9-8ab6-b1921a85a48a.png)
![image](https://user-images.githubusercontent.com/23562932/57613922-55f29f80-7578-11e9-80b9-4fcf6678c0e1.png)
![image](https://user-images.githubusercontent.com/23562932/57613932-5d19ad80-7578-11e9-8150-6b9aad63f269.png)
![image](https://user-images.githubusercontent.com/23562932/57613997-83d7e400-7578-11e9-8195-017c918b3a0c.png)
![image](https://user-images.githubusercontent.com/23562932/57614072-b1249200-7578-11e9-9db5-8a816a312a90.png)
![image](https://user-images.githubusercontent.com/23562932/57614111-c7cae900-7578-11e9-97f6-9e54e345ec69.png)
![image](https://user-images.githubusercontent.com/23562932/57614118-cf8a8d80-7578-11e9-8547-c163773dedac.png)
![image](https://user-images.githubusercontent.com/23562932/57614141-d7e2c880-7578-11e9-87c5-f404c3cae923.png)
![image](https://user-images.githubusercontent.com/23562932/57614158-de714000-7578-11e9-8385-f4dc718160a8.png)
![image](https://user-images.githubusercontent.com/23562932/57614167-e29d5d80-7578-11e9-873a-ffbde142d2df.png)
![image](https://user-images.githubusercontent.com/23562932/57614180-eb8e2f00-7578-11e9-8482-20724a08d89b.png)

Reward open loop log_prob also started to decrease and then increase in the same range:
![image](https://user-images.githubusercontent.com/23562932/57614314-33ad5180-7579-11e9-8eca-0a3766a87788.png)

Other reward statistics doesn't have this clear pattern:
![image](https://user-images.githubusercontent.com/23562932/57614330-41fb6d80-7579-11e9-8f2a-44f973eeb077.png)
![image](https://user-images.githubusercontent.com/23562932/57614344-4a53a880-7579-11e9-8eaf-f2b5f5b1a601.png)

Also in this interval it was scoring less and then it started to score more and more:
![image](https://user-images.githubusercontent.com/23562932/57614379-66efe080-7579-11e9-89b0-34f75f822b72.png)

So can it be that reward model started to learn at the expense of observation model and then both found balance again? Because I don't see much difference in collected dataset around 2M step (although, I only looked at those images of dataset in TensorBoard and it's not much).

# Further work
Now I try with lower free nats in Boxing. Also, I'm working on Freeway and MsPacman. Stay tuned!

**Comment by danijar** [comment]
That's great news! It could be a numerical instability or just that the agent has discovered a new part of the environment that is more surprising. I would suggest to look at the video predictions right before and after this happens (to the degree TensorBoard allows). The reward scales are not very important as I've found. Have you tries disabling overshooting yet?

Besides this, I would recommend a divergence scale that is as high as possible while still allowing for good performance. For example, when you set the divergence scale to zero it could learn to become a deterministic autoencoder which and reconstruct well but is less likely to generalize to state in latent space that the decoder hasn't seen during training.

**Comment by piojanu** [comment]
@danijar thanks for your insights!
1. I didn't try disabling overshooting yet (well, I did once at the very beginning of playing with PlaNet and it seemed to help, but this was before many other tweaks, so this experiment needs to be redone). I can't understand how overshooting might make things worse? It seems like great idea to calculate loss over future predictions (this is what we really care about at the end, to predict future well, not only one step ahead).
2. I've looked into image reconstructions before, during and after this "pit". Before it reconstructions were quite good, then in the pit there is noting but green deck with barriers (?!) and then after the pit reconstructions are amazing. Strange 😮 Second run (I run it more times to confirm results) doesn't have this pit, but it just got to the point when score should start to rise so I don't know just yet if this will work.
3. Yea, I even saw it in my experiments that low divergence scale yields really noisy future predictions. Now I'm running Boxing with `3E-02` and Freeway with `8E-03` so the best parameters from Random Search, but I'll try with higher scale for Freeway too (same as in Boxing).
